README
Bug Reports
===========

Please report any problems to the launchpad bug tracker: https://bugs.launchpad.net/python-crontab

Description
===========

Crontab module for read and writing crontab files and accessing the system cron
automatically and simply using a direct API.

Comparing the below chart to http://en.wikipedia.org/wiki/Cron#CRON_expression
you will note that W, L, # and ? symbols are not supported.

============= =========== ================= ===========================
Field Name    Mandatory   Allowed Values    Allowed Special Characters
============= =========== ================= ===========================
Minutes       Yes         0-59              \* / , -
Hours         Yes         0-23              \* / , -
Day of month  Yes         1-31              \* / , -
Month         Yes         1-12 or JAN-DEC   \* / , -
Day of week   Yes         0-6 or SUN-SAT    \* / , -
============= =========== ================= ===========================

Supported special cases allow crontab lines to not use fields. These are the supported aliases:

=========== ===========
Case        Meaning
=========== ===========
@reboot     Every boot
@hourly     0 * * * *
@daily      0 0 * * *
@weekly     0 0 * * 0
@monthly    0 0 1 * *
@yearly     0 0 1 1 *
@annually   0 0 1 1 *
@midnight   0 0 * * *
=========== ===========

How to Use the Module
=====================

Getting access to a crontab can happen in four ways::

    from crontab import CronTab

    system_cron = CronTab()
    user_cron = CronTab('root')
    file_cron = CronTab(tabfile='filename.tab')
    mem_cron = CronTab(tab="""
      * * * * * command
    """)

Creating a new job is as simple as::

    job  = cron.new(command='/usr/bin/echo')

And setting the job's time restrictions::

    job.minute.during(5,50).every(5)
    job.hour.every(4)
    job.day.on(4, 5, 6)

    job.dow.on('SUN')
    job.month.during('APR', 'NOV')

Creating a job with a comment::

    job = cron.new(command='/foo/bar',comment='SomeID')

Disabled or Enable Job::

    job.enable()
    job.enable(False)
    False == job.is_enabled()

Validity Check::

    True == job.is_valid()

Use a special syntax::

    job.every_reboot()

Find an existing job by command::

    list = cron.find_command('bar')

Find an existing job by comment::

    list = cron.find_comment('ID or some text')

Set and get the comment for a job::

    comment = job.meta(['New Comment for job'])

Clean a job of all rules::

    job.clear()

Iterate through all jobs::

    for job in cron:
        print job

Iterate through all lines::

    for line in cron.lines:
        print line

Remove Items::

    cron.remove( job )
    cron.remove_all('echo')

Write CronTab back to system or filename::

    cron.write()

Write CronTab to new filename::

    cron.write( 'output.tab' )

Log Functionality
=================

The log functionality will read a cron log backwards to find you the last run
instances of your crontab and cron jobs.

The crontab will limit the returned entries to the user the crontab is for.

    cron = CronTab(user='root')

    for d in cron.log:
        print d['pid'] + " - " + d['date']

Each job can return a log iterator too, these are filtered so you can see when
the last execution was.

    for d in cron.find_command('echo')[0].log:
        print d['pid'] + " - " + d['date']

Schedule Functionality
======================

If you have the croniter python module installed, you will have access to a
schedule on each job. For example if you want to know when a job will next run:

    schedule = job.schedule(date_from=datetime.now())

This creates a schedule croniter based on the job from the time specified. The
default date_from is the current date/time if not specified. Next we can get
the datetime of the next job:

    datetime = schedule.get_next()

Or the previous:

    datetime = schedule.get_prev()

The get methods work in the same way as the default croniter, except that they
will return datetime objects by default instead of floats. If you want the
original functionality, pass float into the method when calling:

    datetime = schedule.get_current(float)

If you don't have the croniter module installed, you'll get an ImportError when
you first try using the schedule function on your cron job object.

Extra Support
=============

 - Support for SunOS with compatibility mode
 - Python 3.2 and Python 2.7 tested
 - Windows support works for manual crontabs only


ChangeLog
python-crontab (1.4.1) precise; urgency=low

 * Lots of bug fixes

 -- Martin Owens (DoctorMO) <doctormo@gmail.com>  Thu, 28 May 2013 13:07:02 -0500

python-crontab (1.3.0) precise; urgency=low

 * Add optional croniter support
 * Improvements in testing

 -- Martin Owens (DoctorMO) <doctormo@gmail.com>  Thu, 05 Apr 2013 00:22:58 -0500

python-crontab (1.2.0) precise; urgency=low

  * Add test for enums (MON-SUN/JAN-DEC)
  * Fix problems with enums
  * Make sure enums are preserved
  * Sort output for ranges and parts
  * Allow compatability mode to be set when creating crontab object
  * Allow tabs and tabfiles to be specified for opening

 -- Martin Owens (DoctorMO) <doctormo@gmail.com>  Thu, 19 Sep 2012 12:12:31 -0400

python-crontab (1.1.0) precise; urgency=low

  * Add Python3 support and make sure it all works for both 2.7 and 3.2

 -- Martin Owens (DoctorMO) <doctormo@gmail.com>  Thu, 19 Sep 2012 12:12:31 -0400

python-crontab (1.0.0) precise; urgency=low

  * Add many tests to complete the functionality
  * Improve compatibility mode and test it
  * Change API of slice calls to be properties instead of methods

 -- Martin Owens (DoctorMO) <doctormo@gmail.com>  Thu, 16 Aug 2012 21:39:52 -0400

python-crontab (0.9.7) precise; urgency=low

  * Add unix/sunos compatibility mode

 -- Martin Owens (DoctorMO) <doctormo@gmail.com>  Thu, 14 Aug 2012 21:39:17 -0400

python-crontab (0.9.3) jaunty; urgency=low

  * Fixed problem with crontab line endings

 -- Martin Owens (DoctorMO) <doctormo@gmail.com>  Tue, 18 Sep 2009 06:36:24 -0500
python-crontab (0.9.2) jaunty; urgency=low

  * Fixed problem with comments and invalid line printing

 -- Martin Owens (DoctorMO) <doctormo@gmail.com>  Tue, 06 Jul 2009 20:10:28 -0500
python-crontab (0.9.1) intrepid; urgency=low

  * Fixed an error with @reboot

 -- Martin Owens (DoctorMO) <doctormo@gmail.com>  Tue, 07 Apr 2009 17:39:08 -0500
python-crontab (0.9) hardy; urgency=low

  * Fix lots of pylint errors and warnings.

 -- Martin Owens (DoctorMO) <doctormo@gmail.com>  Sat, 13 Dec 2008 21:32:01 -0800
python-crontab (0.8) hardy; urgency=low

  * Fix a number of problems fixed as reported.

 -- Martin Owens (DoctorMO) <doctormo@gmail.com>  Thu, 09 Oct 2008 20:50:01 -0500
python-crontab (0.7) hardy; urgency=low

  * Fix parsing of special entries
  * Fix looping and printing of crons
  * Updated documentation to reflect new features.

 -- Martin Owens (DoctorMO) <doctormo@gmail.com>  Fri, 23 Aug 2008 00:57:20 -0400
python-crontab (0.6) hardy; urgency=low

  * Fix problems with clearing of cron tasks
  * Don't let speacials mask cleared tasks
  * Add some comments for crontab methods
  * Add remove_all method to crontab

 -- Martin Owens (DoctorMO) <doctormo@gmail.com>  Thu, 20 Aug 2008 19:00:00 -0500
python-crontab (0.5) hardy; urgency=low

  * Package created as base module for other packages.

 -- Martin Owens (DoctorMO) <doctormo@gmail.com>  Thu, 12 Aug 2008 00:15:00 -0500


tests/add.py
#!/usr/bin/python

import sys
sys.path.append('../')

from crontab import CronTab

tab = CronTab()
tab.new(command="/foo/bar")
tab.write()



tests/list.py
#!/usr/bin/python

import sys
sys.path.append('../')

import crontab
from crontab import CronTab

t = CronTab()

for slice in t:
	print slice



tests/remove.py
#!/usr/bin/python

import sys
sys.path.append('../')

import crontab
from crontab import CronTab

t = CronTab()
r = t.remove_all("bar")
t.write()



cronlog.py
#
# Copyright 2013, Martin Owens <doctormo@gmail.com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

import os
import re
import sys
import string

from dateutil import parser as dateparse

MATCHER = r'(?P<date>\w+ +\d+ +\d\d:\d\d:\d\d) (?P<host>\w+) CRON\[(?P<pid>\d+)\]: \((?P<user>\w+)\) CMD \((?P<cmd>.*)\)'

def size(filename):
    return os.stat(filename)[6]

class LogReader(object):
    """Opens a Log file, reading backwards and watching for changes"""
    def __init__(self, filename, mass=4096):
        self.filename = filename
        self.pipe     = open(filename, 'rb')
        self.mass     = mass
        self.size     = -1
        self.read     = -1

    def readlines(self, until=0):
        """Iterator for reading lines from a file backwards"""
        if not self.pipe:
            raise IOError("Can't readline, no opened file.")
        # Always seek to the end of the file, this accounts for file updates
        # that happen during our running process.
        self.size = size(self.filename)
        block_num = 0
        location  = self.size
        halfline  = ''

        while location > until:
            location -= self.mass
            if location < 0:
                location = 0
            self.pipe.seek(location)
            data = unicode(self.pipe.read(self.mass) + halfline).split('\n')
            if location != 0:
                halfline = data.pop(0)
            loc = location + self.mass
            data.reverse()
            for line in data:
                if line.strip() == '':
                    continue
                yield (loc, line)
                loc -= len(line)

    def __iter__(self):
        for (offset, line) in self.readlines():
            yield line


class CronLog(LogReader):
    def __init__(self, filename=None, user=None):
        if not filename:
            filename = '/var/log/syslog'
        LogReader.__init__(self, filename)
        self.user = user

    def for_program(self, command):
        return ProgramLog(self, command)

    def __iter__(self):
        for (offset, line) in self.readlines():
            c = re.match(MATCHER, unicode(line))
            datum = c and c.groupdict()
            if datum and (not self.user or datum['user'] == self.user):
                datum['date'] = dateparse.parse(datum['date'])
                yield datum


class ProgramLog(object):
    """Specific log control for a single command/program"""
    def __init__(self, log, command):
        self.log = log
        self.command = command

    def __iter__(self):
        for entry in self.log:
            if entry['cmd'] == unicode(self.command):
                yield entry
            #else:
            #    print "'%s' != '%s'" % (entry['cmd'], self.command)




crontab.py
#
# Copyright 2013, Martin Owens <doctormo@gmail.com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
# Rewritten from scratch, but based on the code from gnome-schedule by:
# - Philip Van Hoof <me at pvanhoof dot be>
# - Gaute Hope <eg at gaute dot vetsj dot com>
# - Kristof Vansant <de_lupus at pandora dot be>
#
EXAMPLE_USE = """
from crontab import CronTab
import sys

# Create a new non-installed crontab
cron = CronTab(tab='')
job  = cron.new(command='/usr/bin/echo')

job.minute.during(5,50).every(5)
job.hour.every(4)

job.dow.on('SUN')
job.month.during('APR', 'NOV')

job2 = cron.new(command='/foo/bar',comment='SomeID')
job2.every_reboot()

list = cron.find_command('bar')
job3 = list[0]
job3.clear()
job3.minute.every(1)

sys.stdout.write(str(cron.render()))

job3.enable(False)

for job4 in cron.find_command('echo'):
    sys.stdout.write(job4)

for job5 in cron.find_comment('SomeID'):
    sys.stdout.write(job5)

for job6 in cron:
    sys.stdout.write(job5)

cron.remove_all('echo')
cron.remove_all('/foo/bar')
cron.write()

# Croniter Extentions allow you to ask for the scheduled job times, make
# sure you have croniter installed, it's not a hard dependancy.

job3.schedule().get_next()
job3.schedule().get_prev()

"""

import os, re, sys
import tempfile
import subprocess as sp

from datetime import datetime
from cronlog import CronLog

__pkgname__ = 'python-crontab'
__version__ = '1.4.3'

ITEMREX = re.compile('^\s*([^@#\s]+)\s+([^@#\s]+)\s+([^@#\s]+)' +
    '\s+([^@#\s]+)\s+([^@#\s]+)\s+([^#\n]*)(\s+#\s*([^\n]*)|$)')
SPECREX = re.compile('@(\w+)\s([^#\n]*)(\s+#\s*([^\n]*)|$)')
DEVNULL = ">/dev/null 2>&1"

MONTH_ENUM = [ None,
    'jan', 'feb', 'mar', 'apr', 'may',
    'jun', 'jul', 'aug', 'sep', 'oct',
    'nov', 'dec',
]
WEEK_ENUM  = [
    'sun', 'mon', 'tue', 'wed', 'thu',
    'fri', 'sat', 'sun',
]

SPECIALS = {
    "reboot"  : '@reboot',
    "hourly"  : '0 * * * *',
    "daily"   : '0 0 * * *',
    "weekly"  : '0 0 * * 0',
    "monthly" : '0 0 1 * *',
    "yearly"  : '0 0 1 1 *',
    "annually": '0 0 1 1 *',
    "midnight": '0 0 * * *'
}

S_INFO = [
    { 'name' : 'Minutes',      'max_v' : 59, 'min_v' : 0 },
    { 'name' : 'Hours',        'max_v' : 23, 'min_v' : 0 },
    { 'name' : 'Day of Month', 'max_v' : 31, 'min_v' : 1 },
    { 'name' : 'Month',        'max_v' : 12, 'min_v' : 1, 'enum' : MONTH_ENUM },
    { 'name' : 'Day of Week',  'max_v' : 7,  'min_v' : 0, 'enum' : WEEK_ENUM },
]

# Detect Python3 and which OS for temperments.
import platform
py3 = platform.python_version()[0] == '3'
WinOS = platform.system() == 'Windows'
SunOS = not WinOS and (os.uname()[0] == "SunOS" or os.getenv('SunOS_TEST'))
CRONCMD = "/usr/bin/crontab"
if sys.argv[0].startswith('test_'):
    CRONCMD = './data/crontest'

if py3:
    unicode = str
    basestring = str

try:
    # Croniter is an optional import
    from croniter.croniter import croniter as _croniter
    class croniter(_croniter):
        """Same as normal croniter, but always return datetime objects"""
        def get_next(self, type_ref=datetime):
            return _croniter.get_next(self, type_ref)
        def get_prev(self, type_ref=datetime):
            return _croniter.get_prev(self, type_ref)
        def get_current(self, type_ref=datetime):
            return _croniter.get_current(self, type_ref)
except ImportError:
    croniter = None


class CronTab(object):
    """
    Crontab object which can access any time based cron using the standard.

    user    - Set the user of the crontab (defaults to $USER)
    tab     - Use a string variable as the crontab instead of installed crontab
    tabfile - Use a file for the crontab instead of installed crontab
    log     - Filename for logfile instead of /var/log/syslog

    """
    def __init__(self, user=None, tab=None, tabfile=None, log=None):
        self.lines = None
        self.crons = None
        self.filen = None
        # Protect windows users
        self.root  = not WinOS and os.getuid() == 0
        self.user  = user
        # Detect older unixes and help them out.
        self.intab = tab
        self.read(tabfile)
        self._log = log

    @property
    def log(self):
        if self._log == None or isinstance(self._log, basestring):
            self._log = CronLog(self._log, user=self.user or 'root')
        return self._log

    def read(self, filename=None):
        """
        Read in the crontab from the system into the object, called
        automatically when listing or using the object. use for refresh.
        """
        self.crons = []
        self.lines = []
        if self.intab != None:
            lines = self.intab.split('\n')
        elif filename:
            self.filen = filename
            with open(filename, 'r') as fhl:
                lines = fhl.readlines()
        else:
            p = sp.Popen(self._read_execute(), stdout=sp.PIPE, stderr=sp.PIPE)
            (out, err) = p.communicate()
            lines = out.decode('utf-8').split("\n")
        for line in lines:
            cron = CronItem(line, cron=self)
            if cron.is_valid():
                self.crons.append(cron)
                self.lines.append(cron)
            else:
                self.lines.append(line.replace('\n',''))

    def write(self, filename=None):
        """Write the crontab to the system. Saves all information."""
        if filename:
            self.filen = filename

        # Add to either the crontab or the internal tab.
        if self.intab != None:
          self.intab = self.render()
          # And that's it if we never saved to a file
          if not self.filen:
              return

        if self.filen:
            fileh = open(self.filen, 'w')
        else:
            filed, path = tempfile.mkstemp()
            fileh = os.fdopen(filed, 'w')

        fileh.write(self.render())
        fileh.close()

        if not self.filen:
            # Add the entire crontab back to the user crontab
            sp.Popen(self._write_execute(path)).wait()
            os.unlink(path)

    def render(self):
        """Render this crontab as it would be in the crontab."""
        crons = []
        for cron in self.lines:
            crons.append(unicode(cron))
        result = '\n'.join(crons)
        if result and result[-1] not in [ '\n', '\r' ]:
            result += '\n'
        return result

    def new(self, command='', comment=''):
        """
        Create a new cron with a command and comment.

        Returns the new CronItem object.
        """
        item = CronItem(command=command, meta=comment, cron=self)
        self.crons.append(item)
        self.lines.append(item)
        return item

    def find_command(self, command):
        """Return a list of crons using a command."""
        result = []
        for cron in self.crons:
            if cron.command.match(command):
                result.append(cron)
        return result

    def find_comment(self, comment):
        """Return a list of crons using the comment field."""
        result = []
        for cron in self.crons:
            if cron.meta() == comment:
                result.append(cron)
        return result

    def remove_all(self, command):
        """Removes all crons using the stated command."""
        l_value = self.find_command(command)
        for c_value in l_value:
            self.remove(c_value)

    def remove(self, item):
        """Remove a selected cron from the crontab."""
        # The last item often has a trailing line feed
        if self.crons[-1] == item and self.lines[-1] == '':
            self.lines.remove(self.lines[-1])
        self.crons.remove(item)
        self.lines.remove(item)

    def _read_execute(self):
        """Returns the command line for reading a crontab"""
        return [ CRONCMD, '-l' ] + self._user_execute()

    def _write_execute(self, path):
        """Return the command line for writing a crontab"""
        return [ CRONCMD, path ] + self._user_execute()

    def _user_execute(self):
        """User command switches to append to the read and write commands."""
        if self.user:
            return [ '-u', str(self.user) ]
        return []

    def __iter__(self):
        return self.crons.__iter__()

    def __unicode__(self):
        return self.render()

    def __str__(self):
        return self.render()


class CronItem(object):
    """
    An item which objectifies a single line of a crontab and
    May be considered to be a cron job object.
    """
    def __init__(self, line=None, command='', meta='', cron=None):
        self.valid = False
        self.enabled = True
        self.slices  = []
        self.special = False
        self.cron  = cron

        self._meta   = meta
        self._log    = None

        self.set_slices()
        if line and line.strip():
            self.parse(line.strip())
        elif command:
            self.command = CronCommand(unicode(command))
            self.valid = True

    def delete(self):
        if not self.cron:
            sys.stderr.write("Cron item is not associated with a crontab!\n")
        else:
            self.cron.remove(self)

    def parse(self, line):
        """Parse a cron line string and save the info as the objects."""
        if not line or line[0] == '#':
            self.enabled = False
            line = line[1:].strip()
        result = ITEMREX.findall(line)
        if result:
            o_value = result[0]
            self.command = CronCommand(o_value[5])
            self._meta   = o_value[7]
            try:
                self.set_slices( o_value )
            except KeyError:
                self.enabled = False
            else:
                self.valid = True
        elif line.find('@') < line.find('#') or line.find('#')==-1:

            result = SPECREX.findall(line)
            if result and result[0][0] in SPECIALS:
                o_value = result[0]
                self.command = CronCommand(o_value[1])
                self._meta   = o_value[3]
                value = SPECIALS[o_value[0]]
                if value.find('@') != -1:
                    self.special = value
                else:
                    self.set_slices( value.split(' ') )
                self.valid = True

    def set_slices(self, o_value=None):
        """Set the values of this slice set"""
        self.slices = []
        for i_value in range(0, 5):
            if not o_value:
                o_value = [None, None, None, None, None]
            self.slices.append(
                CronSlice(job=self, value=o_value[i_value],
                    **S_INFO[i_value]))

    def enable(self, enabled=True):
        """Set if this cron job is enabled or not"""
        if enabled in [True, False]:
            self.enabled = enabled
        return self.enabled

    def is_enabled(self):
        """Return true if this job is enabled (not commented out)"""
        return self.enabled

    def is_valid(self):
        """Return true if this job is valid"""
        return self.valid

    def render_time(self):
        return ' '.join([ unicode(self.slices[i]) for i in range(0, 5) ])

    def render(self):
        """Render this set cron-job to a string"""
        time = self.render_time()

        if self.special or time in SPECIALS.values():
            if self.special:
                time = self.special
            else:
                for (name, value) in SPECIALS.items():
                    if value == time:
                        time = "@%s" % name

        result = "%s %s" % (time, unicode(self.command))
        if self.meta():
            result += " # " + self.meta()
        if not self.enabled:
            result = "# " + result
        return result

    def meta(self, value=None):
        """Return or set the meta value to replace the set values"""
        if value:
            self._meta = value
        return self._meta

    def every_reboot(self):
        """Set to every reboot instead of a time pattern"""
        self.special = '@reboot'

    def clear(self):
        """Clear the special and set values"""
        self.special = None
        for slice_v in self.slices:
            slice_v.clear()

    def schedule(self, date_from=None):
        """Return a croniter schedule is available."""
        if not date_from:
            date_from = datetime.now()
        if croniter:
            return croniter(self.render_time(), date_from)
        raise ImportError("Croniter is not available. Please install croniter"+\
                         " python module via pip or your package manager")

    @property
    def log(self):
        """Return a cron log specific for this job only"""
        if not self._log and self.cron:
            self._log = self.cron.log.for_program(self.command)
        return self._log

    @property
    def minute(self):
        """Return the minute slice"""
        return self.slices[0]

    @property
    def hour(self):
        """Return the hour slice"""
        return self.slices[1]

    @property
    def dom(self):
        """Return the day-of-the month slice"""
        return self.slices[2]

    @property
    def month(self):
        """Return the month slice"""
        return self.slices[3]

    @property
    def dow(self):
        """Return the day of the week slice"""
        return self.slices[4]

    def __repr__(self):
        return "<CronJob '%s'>" % str(self)

    def __eq__(self, value):
        return str(self) == str(value)

    def __str__(self):
        return self.__unicode__()

    def __unicode__(self):
        if not self.is_valid():
            sys.stderr.write("Ignoring invalid crontab line\n")
            return "# " + unicode(self.render())
        return self.render()


class CronSlice(object):
    """Cron slice object which shows a time pattern"""
    def __init__(self, name, min_v, max_v, enum=None, value=None, job=None):
        self.name  = name
        self.min   = min_v
        self.max   = max_v
        self.job   = job
        self.enum  = enum
        self.parts = []
        if value:
            self._set_value(value)

    def _set_value(self, value):
        """Set values into the slice."""
        self.parts = []
        for part in value.split(','):
            if part.find("/") > 0 or part.find("-") > 0 or part == '*':
                self.parts.append( self.get_range( part ) )
            else:
                try:
                    self.parts.append( self._v(part) )
                except ValueError:
                    raise ValueError('Unknown cron time part for %s: %s' % (
                        self.name, part))

    def render(self, resolve=False):
        """Return the slice rendered as a crontab.

        resolve - return integer values instead of enums (default False)

        """
        if len(self.parts) == 0:
            return '*'
        return _render_values(self.parts, ',', resolve)

    def __repr__(self):
        return "<CronSlice '%s'>" % str(self)

    def __eq__(self, value):
        return str(self) == str(value)

    def __str__(self):
        return self.__unicode__()

    def __unicode__(self):
        return self.render()

    def every(self, n_value):
        """Set the every X units value"""
        self.parts = [ self.get_range( int(n_value) ) ]
        return self.parts[0]

    def on(self, *n_value):
        """Set the on the time value."""
        for av in n_value:
            self.parts += self._v(av),

    def during(self, vfrom, vto):
        """Set the During value, which sets a range"""
        self.parts.append(self.get_range(self._v(vfrom), self._v(vto)))
        return self.parts[-1]

    def clear(self):
        """clear the slice ready for new vaues"""
        self.parts = []

    def get_range(self, *vrange):
        """Return a cron range for this slice"""
        return CronRange( self, *vrange )

    def _v(self, v):
        """Support wrapper for enumerations and check for range"""
        try:
            out = get_cronvalue(v, self.enum)
        except ValueError:
            raise ValueError("Unrecognised '%s'='%s'" % (self.name, v))
        except KeyError:
            raise KeyError("No enumeration '%s' got '%s'" % (self.name, v))

        if int(out) < self.min and int(out) > self.max:
            raise ValueError("Invalid value '%s', expected %d-%d for %s" % (
                str(v), self.min, self.max, self.name))
        return out


def get_cronvalue(value, enums):
    """Returns a value as int (pass-through) or a special enum value"""
    if isinstance(value, int):
        return value
    elif str(value).isdigit():
        return int(str(value))
    if not enums:
        raise KeyError("No enumeration allowed")
    return CronValue(str(value), enums)


class CronValue(object):
    """Represent a special value in the cron line"""
    def __init__(self, value, enums):
        self.enum = value
        self.value = enums.index(value.lower())

    def __lt__(self, value):
        return self.value < int(value)
    def __repr__(self):
        return str(self)
    def __str__(self):
        return self.enum
    def __int__(self):
        return self.value


def _render_values(values, sep=',', resolve=False):
    """Returns a rendered list, sorted and optionally resolved"""
    if len(values) > 1:
        values.sort()
    return sep.join([ _render(val, resolve) for val in values ])

def _render(value, resolve=False):
    """Return a single value rendered"""
    if isinstance(value, CronRange):
        return value.render(resolve)
    if resolve:
        return str(int(value))
    return str(value)



class CronRange(object):
    """A range between one value and another for a time range."""
    def __init__(self, vslice, *vrange):
        self.slice = vslice
        self.seq   = 1
        self.cron  = None

        if not vrange:
            self.all()
        elif isinstance(vrange[0], basestring):
            self.parse(vrange[0])
        elif isinstance(vrange[0], int) or isinstance(vrange[0], CronValue):
            if len(vrange) == 2:
                (self.vfrom, self.vto) = vrange
            else:
                self.seq = vrange[0]
                self.all()

    def parse(self, value):
        """Parse a ranged value in a cronjob"""
        if value.find('/') > 0:
            value, seq = value.split('/')
            self.seq = int(seq)
        if value.find('-') > 0:
            vfrom, vto = value.split('-')
            self.vfrom = self.slice._v(vfrom)
            self.vto  = self.slice._v(vto)
        elif value == '*':
            self.all()
        else:
            raise ValueError('Unknown cron range value %s' % value)

    def all(self):
        """Set this slice to all units between the miniumum and maximum"""
        self.vfrom = self.slice.min
        self.vto  = self.slice.max

    def render(self, resolve=False):
        """Render the ranged value for a cronjob"""
        value = '*'
        if int(self.vfrom) > self.slice.min or int(self.vto) < self.slice.max:
            value = _render_values([self.vfrom, self.vto], '-', resolve)
        if self.seq != 1:
            value += "/%d" % self.seq
        if value != '*' and SunOS:
            value = ','.join(map(str, range(self.vfrom, self.vto+1, self.seq)))
        return value

    def every(self, value):
        """Set the sequence value for this range."""
        self.seq = int(value)

    def __lt__(self, value):
        return int(self.vfrom) < int(value)

    def __int__(self):
        return int(self.vfrom)

    def __str__(self):
        return self.__unicode__()

    def __unicode__(self):
        return self.render()


class CronCommand(object):
    """Represent a cron command as an object."""
    def __init__(self, line):
        self._command = line

    def match(self, command):
        """Match the command given"""
        return command in self._command

    def command(self):
        """Return the command line"""
        return self._command

    def __str__(self):
        """Return a string as a value"""
        # TODO: encode with the system's default encoding?
        # i.e. locale.getpreferredencoding()
        return self.__unicode__()

    def __unicode__(self):
        """Return unicode command line value"""
        return self.command().strip()

